## Differential Accessibility Analysis of ATAC-seq data

## Introduction

The ATAC-seq data used in this workflow was generated as part of a study that investigated the regulatory role of short-chain fatty acids (SCFAs) propionate and butyrate in the context of colorectal cancer (CRC). Briefly, epigenomic and transcriptomic profiles were compared in SCFA-treated and untreated CRC cells, normal cells and in mouse intestines, with the goal of gaining insight into changes in accessibility and expression of CRC-relevant genes.

The data set used in this workflow comes from propionyl ATAC-seq in CRC cells. The untreataed condition is used as a control. The data identifies genomic regions showing increased transposase accessibility following treatment (10mM) with sodium propionate (NaPr). 

Once `diffbind.sh` has been run and the resulting `.RData` object and `report.csv` file are ready to be loaded into `R`, `BiocManager`, `DiffBind`, `EnsDb.Hsapiens.v86`, `AnnotationDbi2`, `org.Hs.eg.db`, `ChIPpeakAnno`, `GenomicRanges`, packages will first have to be installed.

Once the `.RData` object has been loaded and the dependencies installed, `DiffBind` will be used for exploratory analysis and for identifying differentially bound genomic coordinates. `GenomicRanges` will then be used to convert genomic coordinates to an annotated object. `ChIPpeakAnno` will be used for annotation of differentially bound regions, and `ChIPseeker` will provide average profile and heatmap of peak accessibility to TSS regions and genomic feature distribution. Finally, `ReactomePA` will be used to perform pathway analysis. 

## Goals

This workflow will perform the following analyses:

-   Compare correlations between replicates and conditions
-   Compare differential accessibility between histone marks
-   Display fold change of normalized read counts
-   Identify differentially bound regions
-   Create TSS distribution profiles
-   Annotate peaks and show feature distribution profiles
-   Perform pathway analysis

## Setup

This R Markdown (.Rmd) file will provide a walk-through of exploratory data analysis followed by differential accessibility analysis, annotation and pathway analysis of propionyl ATAC-seq data. Each `Rmd` file has an initial `setup` chunk that applies some settings to all of the code chunks.

```{css, echo = FALSE}
pre, code {white-space:pre !important; overflow-x:auto !important}
```


```{r setup, include=FALSE}
# when you "knit" this file, do you want the resulting PDF to print the code in each chunk (TRUE = yes)?
knitr::opts_chunk$set(echo = TRUE)

################################################################################
# set your working directory
####CHANGE THIS TO THE APPROPRIATE PATH
knitr::opts_knit$set(root.dir = '~/Desktop/Data/ATAC-seq/')
################################################################################
# note that outside of an Rmd code chunk, use `setwd()` to set the working directory in R
```

As in a regular R script in RStudio, a single line of code can be run with Command-Enter (Mac OS) or Ctrl-Enter (Windows). Whole chunks of code can be run with Command(/Ctrl) + Shift + Enter **or** by clicking the green "\>" button in the top-right corner of the chunk. Alternatively, these options can also be implemented by selecting lines of code and choosing the desired option from the drop-down menu in the "Run" tab, in the top-right corner of the of Source section of RStudio.

To comment or uncomment a series of lines, highlight the lines and use Command(/Ctrl) + Shift + C.

The following `Bioconductor` `R` packages are required for this workflow.
They can be installed using the following commands:

```{r install packages, eval=F}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("EnsDb.Hsapiens.v86")
install.packages("AnnotationDbi")
install.packages("org.Hs.eg.db")
install.packages("ChIPpeakAnno")
install.packages("GenomicRanges")

# If install.packages("packagename") command fails in newer versions of R uncomment and run the commands:
# if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
# BiocManager::install("packagename")
```

Once installed, load the required packages.

```{r}
library(EnsDb.Hsapiens.v86)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(GenomicRanges)
library(DiffBind)
library(ChIPpeakAnno)
```

## Exploratory and Differential Accessibility Analysis

Before annotation and pathway analysis it good to first visualize the data and perform quality control checks. This can be done by loading the `.RData` object generated by `Diffbind` using `diffbind_chip.R` script provided in the repository. 

```{r load data file}
load("~/Desktop/Data/ATAC-seq/prop_atac.RData")
```

The `atac.RData` file should contain the `diff` object, which includes differential accessibility and contrast/covariate information. Make sure they are loaded into your global environment.

Check data frame information: 
```{r display data frame}
# This will display experimental variables defined in the sampleSheet and differentially bound sites identified by DESeq2.
diff
```

Next, it's good to visualize the correlation between different replicates within each condition.

```{r correlation matrix}
# "dba" is function that is built into DiffBind package and can be used to make plots from the "diff" object.
# "Contrast" refers to the differential analysis by factor (or condition). 
# In this example, the contrast is between propionate treated vs untreated groups defined in the sampleSheet.
# Replicates from the same condition should have a higher correlation score than those from different conditions.

dba.plotHeatmap(diff, contrast=1)

# Alternatively, you could also run:
#plot(diff, contrast=1)
```

```{r correlation heatmap}
# Generate a accessibility affinity heatmap showing affinities for differentially bound sites.
dba.plotHeatmap(diff, contrast=1, correlations=FALSE)
```


```{r MA plot}
# Make an MA plot showing differentially bound sites associated with the two histone marks.
dba.plotMA(diff, bFlip=TRUE)

# Sites preferentially associated with Kpr above a significance threshold are shown as red dots in the upper panel. 
```


```{r volcano plot}
# Make a volcano plot showing the log2 fold change vs -log10 FDR-adjusted (<0.05) p-value.
dba.plotVolcano(diff, bFlip=TRUE)
```


## Annotation of Differentially Bound Regions

```{r annotation data}
# Create an annotation data object.
annoData <- annoGR(EnsDb.Hsapiens.v86, feature="gene")
info(annoData)
annoData[1:2]
```

First, load the `report.csv` file generated by `DiffBind` and create a new object, in this case *prop*. 
Make sure the data frame loaded correctly and inspect the top rows by using `head(object)`.

```{r read counts}
prop <- read.csv("prop_atac_report.csv", header=TRUE)
head(prop)
```

Next, make a `GRanges` data frame from report table:
```{r make data frame}
propgr <- makeGRangesFromDataFrame(prop, keep.extra.columns=TRUE)
propgr
```

Annotate to within +/- 1000 bp of transcription start sites (TSS) of a transcript.
```{r annotate data}
prop.anno <- annotatePeakInBatch(propgr, AnnotationData=annoData, output="both", maxgap=1000L, select="all")
```

Annotate by feature.
```{r annotate by feature}
prop.anno$gene_name <- annoData$gene_name[match(prop.anno$feature, names(annoData))]
head(prop.anno$feature)
```

Annotate by gene symbol.
```{r annotate by symbol}
prop.anno$symbol <- mapIds(org.Hs.eg.db, 
                     keys=prop.anno$feature,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
prop.anno
```

Annotate by gene entrezID.
```{r annotate by geneID}
prop.anno$entrezID <- mapIds(org.Hs.eg.db, 
                     keys=prop.anno$feature,
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")
prop.anno
```

Next, make a report of the annotated output.
```{r create data frome}
# write annotated output
diffprop <- data.frame(seqnames=seqnames(prop.anno),
  starts=(start(prop.anno)-1),
  ends=end(prop.anno),
  ENSG=prop.anno$feature,
  ENSG.strand=prop.anno$feature_strand, 
  insideFeature=prop.anno$insideFeature, 
  distancetoFeature=prop.anno$distancetoFeature,
  shortestDistance=prop.anno$shortestDistance,
  symbol=prop.anno$symbol, 
  entrezID=prop.anno$entrezID,
  fold=prop.anno$Fold,
  conc=prop.anno$Conc,
  conc_0mM=prop.anno$Conc_0mM,
  conc_10mM=prop.anno$Conc_10mM,
  FDR=prop.anno$FDR)
head(diffprop)
```

Save the annotated output data table.
```{r write a table}
write.table(diffprop, file="prop.anno.bed", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
dim(diffprop)
```

**NOTE:** The final output contains a table with features showing positive and negative fold change. In this example, the positive vs negative fold change relationship is reversed, with NaPr treatment associated genes labeled as negative fold change. You may wish to select only those regions showing positive or negative fold change (or increased vs decreased accessibility) by opening the `.report.csv` or `.anno.bed` files in MS Excel, selecting the "fold" column, clicking on "Sort & Filter" and selecting only "Greater Than" or "Less Than" "0" entries.

## Profile of Differential Accessibility to TSS Regions

The next several modules will require `ChIPseeker` and `TxDb.Hsapiens.UCSC.hg38.knownGene` packages.

```{r}
# Install `ChIPseeker` package:
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("ChIPseeker")
```


```{r}
# Install `TxDb.Hsapiens.UCSC.hg38.knownGene` package:
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")
```


Load the required packages:
```{r}
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
```

Create a *TxDb* object based on the UCSC hg38 assembly of the human genome.
```{r create annotation object}
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
```

Annotate peaks to 3000 bp upstream and downstream of promoters and generate heatmap.
```{r plot TSS heatmap}
promoter <- getPromoters(TxDb=txdb, upstream=3000, downstream=3000)
tagMatrix <- getTagMatrix(propgr, windows=promoter)
tagHeatmap(tagMatrix, xlim=c(-3000, 3000), color="red")
```

Plot peak distributions profile 3000 bp upstream and downstream of TSS.
```{r plot TSS profile}
plotAvgProf(tagMatrix, xlim=c(-3000, 3000), xlab="Genomic Region (5'->3')", ylab = "Read Count Frequency")
```

Resample peak distribution profile 1000 times by running bootstrapping with 0.95 confidence cutoff.
```{r plot TSS profile and resample}
plotAvgProf(tagMatrix, xlim=c(-3000, 3000), conf = 0.95, resample = 1000)
```

Annotate again to 3000 bp upstream and downstream of TSS by using `org.Hs.eg.db`.
```{r annotate data to TSS}
peakAnno <- annotatePeak(propgr, tssRegion=c(-3000, 3000), TxDb=txdb, annoDb="org.Hs.eg.db")
```

Plot differentially bound feature distribution as a pie or a bar plot.
```{r plot feature distribution}
plotAnnoPie(peakAnno)
plotAnnoBar(peakAnno)
```

## Pathway Analysis of Differentially Accessible Genes

**NOTE:** As mentioned earlier, in this example, the positive vs negative fold change relationship is reversed, with propionate treatment associated genes assigned negative fold change. Pathways that go up or show enrichment are defined as fold < 0 and vise versa.

```{r pathway analysis}
# do pathway analysis in Reactome
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("ReactomePA")
library(ReactomePA)
head(diffprop$entrezID)
```

Check number of features showing increased enrichment or association with NaPr treatment.
```{r number of pathways up}
diffup <- diffprop[which(diffprop$fold < 0), ]
dim(diffup)
```

Check number of features showing decreased enrichment or association with NaPr treatment.
```{r number of pathways down}
diffdown <- diffprop[which(diffprop$fold > 0), ]
dim(diffdown)
```

Identify enriched pathways with p-adjusted value cutoff of 0.05. 
```{r pathway enrichment}
de <- (diffprop$entrezID)
length(de)
de <- de[!is.na(de)]
head(de)
length(de)
x <- enrichPathway(gene=de,pvalueCutoff=0.05, readable=TRUE)
```


Make a dotplot showing pathways that go up or are associated with NaPr treatment.
```{r pathways up}
deup <- diffup$entrezID
deup <- deup[!is.na(deup)]
length(deup)
xup <- enrichPathway(gene=deup,pvalueCutoff=0.05, readable=TRUE)
dotplot(xup, showCategory=10)
```

